### PR TITLE
Update to Nextcloud 16

### LINKED ISSUE #3: Use in Nextcloud 15?
Hi Roeland,
is your app still in development and is it possible to use your app in Nextcloud 15 too?
Thanks a lot 
Michael

### LINKED ISSUE #4: Remove TOTP
There is no need for TOTP here. We can just generate an X char secret.

### PR DESCRIPTION
**Pull request by Flow86**
I've reworked the Plugin ( with some help from https://github.com/nextcloud/twofactor_gateway ) so it fully works with the latest nextcloud server version (=16).

It also does not use a TOTP anymore, since that was completely over the top (see #4)

Most things of #2 are also implemented. Only the "Listen to e-mail changes & Act accordingly" is not implemented. It will always use the primary users email. It currently does not compare it to the one activated or even remember the one used to activate the two-factor authentication.

closes #3 (implemented)
closes #4 (TOTP removed)

### CODE REVIEW

**Comment by RuudschMaHinda** [comment]
Hoping this gets reviewed and passes soon.

File: js/components/EmailInstructions.vue [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<template>
+	<div>
+		<p>
+			The server can send authentication codes to your email address.
```
Comment (ChristophWurst) [inline js/components/EmailInstructions.vue]: missing translation

File: js/components/GatewaySettings.vue [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+					});
+			},
+
+			disable: function () {
+				this.loading = true;
+
+				disable()
+					.then(res => {
+						this.state = res.state;
+                        this.emailAddress = res.emailAddress;
```
Comment (ChristophWurst) [inline js/components/GatewaySettings.vue]: indentation is off

File: lib/AppInfo/Application.php [hunk 3]
```diff
@@ -24,31 +24,14 @@
...
-	const APPNAME='twofactor_email';
+	const APP_NAME = 'twofactor_email';
 
-	public function __construct() {
-		parent::__construct(self::APPNAME);
-
-		$container = $this->getContainer();
-
-		$dispatcher = $container->getServer()->getEventDispatcher();
-		$dispatcher->addListener(StateChanged::class, function (StateChanged $event) use ($container) {
```
Comment (ChristophWurst) [inline lib/AppInfo/Application.php]: why was this removed?

File: lib/Provider/Email.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+use OCP\Security\ISecureRandom;
+use OCP\Template;
+
+class Email implements IProvider, IProvidesIcons, IProvidesPersonalSettings {
+
+	const STATE_DISABLED = 0;
+	const STATE_VERIFYING = 1;
+	const STATE_ENABLED = 2;
+
+    /** @var EmailService */
```
Comment (ChristophWurst) [inline lib/Provider/Email.php]: indentation is off

File: lib/Provider/Email.php [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	protected $session;
+
+	/** @var ISecureRandom */
+	protected $secureRandom;
+
+	/** @var IL10N */
+	protected $l10n;
+
+	public function __construct(EmailService $emailService,
+                                StateStorage $stateStorage,
```
Comment (ChristophWurst) [inline lib/Provider/Email.php]: indentation

File: lib/AppInfo/Application.php (hunk 3 above)
Comment (Flow86) [reply lib/AppInfo/Application.php]: since its not necessary? The provider is enabled somewhere else (in lib/Service/SetupService.php )

File: js/components/EmailInstructions.vue (hunk 1 above)
Comment (Flow86) [reply js/components/EmailInstructions.vue]: added translation tag in 3743fab

File: js/components/GatewaySettings.vue (hunk 2 above)
Comment (Flow86) [reply js/components/GatewaySettings.vue]: fixed in 3743fab

File: lib/Provider/Email.php (hunk 4 above)
Comment (Flow86) [reply lib/Provider/Email.php]: fixed in 3743fab

File: lib/Provider/Email.php (hunk 5 above)
Comment (Flow86) [reply lib/Provider/Email.php]: fixed in 3743fab

File: lib/AppInfo/Application.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/AppInfo/Application.php]: If `StateChangeRegistryUpdater` is obsolete then please remove it

Comment (Flow86) [reply lib/AppInfo/Application.php]: it is removed?

**Comment by Flow86** [comment]
> check the indentation here on github, it does not look right wherever you've changed code :)

Thanks. Stupid editor went crazy again and switched to spaces instead of tabs.

**Comment by rullzer** [comment]
@Flow86 wow thanks a lot for all the work.
I'll have a look soon and see if we can get it in :)

Cool stuff.

**Comment by Flow86** [comment]
@rullzer if you need any additional changes, let me know :-)

File: package.json [hunk 6]
```diff
@@ -4,26 +4,28 @@
...
+  "main": "index.js",
   "directories": {
     "lib": "lib",
     "test": "tests"
   },
   "scripts": {
-    "dev": "webpack --config webpack.dev.js",
-    "watch": "webpack --progress --watch --config webpack.dev.js",
-    "build": "webpack --progress --hide-modules --config webpack.prod.js",
+    "build": "./node_modules/webpack-cli/bin/cli.js --config js/webpack.prod.config.js",
```
Comment (rullzer) [inline package.json]: It seems those files are missing

File: package.json [hunk 7]
```diff
@@ -4,26 +4,28 @@
...
+    "dev": "./node_modules/webpack-cli/bin/cli.js --config js/webpack.dev.config.js --watch",
     "test": "echo \"Error: no test specified\" && exit 1"
   },
   "dependencies": {
-    "axios": "^0.18.0",
-    "vue": "^2.5.17"
+    "css-loader": "^2.1.1",
+    "jquery": "^3.4.1",
+    "nextcloud_fetch": "^2.1.0",
```
Comment (rullzer) [inline package.json]: We kind of defaulted to nextcloud-axios for network stuff. Also since @ChristophWurst is not maintaining fetch right now. And this will make it do :boom: on older browsers.

File: Makefile [hunk 8]
```diff
@@ -1,4 +1,4 @@
-app_name=bruteforcesettings
+app_name=twofactor_email
```
Comment (rullzer) [inline Makefile]: Nothing for now. But we should move this to krankerl as well @ChristophWurst

File: appinfo/app.php [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<?php
+
+include_once __DIR__ . '/../vendor/autoload.php';
```
Comment (rullzer) [inline appinfo/app.php]: I think this can die as well. But I'll check later.

File: composer.json [hunk 10]
```diff
@@ -1,7 +1,4 @@
 {
-  "require": {
-    "christian-riesen/otp": "1.*"
```
Comment (rullzer) [inline composer.json]: Yep much better to kill this :rocket:

File: js/components/EmailInstructions.vue [hunk 11]
```diff
@@ -0,0 +1,17 @@
+<template>
```
Comment (rullzer) [inline js/components/EmailInstructions.vue]: So small nitpicking in general. The convention we came up with is having a src folder. Where all the js srces are.
And then we have build it into `js`.

But I could tackle that in a later PR.

File: js/service/registration.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
+import {nc_fetch_json} from 'nextcloud_fetch';
+
+export function getVerificationState () {
+	let url = OC.generateUrl('/apps/twofactor_email/settings/state')
```
Comment (rullzer) [inline js/service/registration.js]: After this gets merged @ChristophWurst can confert this into his fancy tiny package stuff

File: package.json (hunk 6 above)
Comment (rullzer) [reply package.json]: Nah my bad

File: package.json (hunk 7 above)
Comment (ChristophWurst) [reply package.json]: it's even deprecated via npm. if you install it it should recommend nextcloud-axios instead
